arangod/Aql/QueryRegistry.cpp

```diff
@@ -182,146 +188,206 @@ void* QueryRegistry::openEngine(EngineId id, EngineType type) {
 void QueryRegistry::closeEngine(EngineId engineId) {
   LOG_TOPIC("3f0c9", DEBUG, arangodb::Logger::AQL)
       << "returning engine with id " << engineId;
-  WRITE_LOCKER(writeLocker, _lock);
 
-  auto it = _engines.find(engineId);
-  if (it == _engines.end()) {
-    LOG_TOPIC("97351", DEBUG, arangodb::Logger::AQL)
-        << "Found no engine with id " << engineId;
-    return;
-  }
+  // finishing the query will call back into the registry to destroy the query,
+  // so this must be done outside the lock
+  futures::Promise<std::shared_ptr<ClusterQuery>> finishPromise;
+  std::shared_ptr<ClusterQuery> queryToFinish;
 
-  EngineInfo& ei = it->second;
-  if (!ei._isOpen) {
-    LOG_TOPIC("45b8e", DEBUG, arangodb::Logger::AQL)
-        << "engine id " << engineId << " was not open.";
-    THROW_ARANGO_EXCEPTION_MESSAGE(
-        TRI_ERROR_INTERNAL, "engine with given vocbase and id is not open");
-  }
+  {
+    WRITE_LOCKER(writeLocker, _lock);
 
-  ei._isOpen = false;
+    auto it = _engines.find(engineId);
+    if (it == _engines.end()) {
+      LOG_TOPIC("97351", DEBUG, arangodb::Logger::AQL)
+          << "Found no engine with id " << engineId;
+      return;
+    }
 
-  if (ei._queryInfo) {
-    TRI_ASSERT(ei._queryInfo->_numOpen > 0);
-    ei._queryInfo->_numOpen--;
-    if (!ei._queryInfo->_query->killed() && ei._queryInfo->_expires != 0) {
-      ei._queryInfo->_expires = TRI_microtime() + ei._queryInfo->_timeToLive;
+    EngineInfo& ei = it->second;
+    if (!ei._isOpen) {
+      LOG_TOPIC("45b8e", DEBUG, arangodb::Logger::AQL)
+          << "engine id " << engineId << " was not open.";
+      THROW_ARANGO_EXCEPTION_MESSAGE(
+          TRI_ERROR_INTERNAL, "engine with given vocbase and id is not open");
     }
-    LOG_TOPIC("5ecdc", TRACE, arangodb::Logger::AQL)
-        << "closing engine " << engineId
-        << ", query id: " << ei._queryInfo->_query->id()
-        << ", numOpen: " << ei._queryInfo->_numOpen;
-  } else {
-    LOG_TOPIC("ae981", TRACE, arangodb::Logger::AQL)
-        << "closing engine " << engineId << ", no query";
+
+    ei._isOpen = false;
```

Comment:
Unnecessary assign.

Reply:
Why do you think this is unnecessary? This way we mark the engine as unused, which is actually quite important.

Reply:
I think it actually is necessary.
We enter this code with _isOpen == true (otherwise we throw in the lines above)
And we need to leave it with _isOpen == false as we are giving back ownership to the Registry.

Reply:
if (cond == false) throw
// Here cond is always true